File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/minimum-window-substring/code/cpp/minimum_window_substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.size() < t.size()) return "";
        unordered_map<char, int> mp;
        for (char c : t) mp[c]++;
        int left = 0, minLen = INT_MAX, minStart = 0, count = t.size();
        for (int right = 0; right < s.size(); right++) {
            if (mp.count(s[right])) {
                if (mp[s[right]] > 0) count--;
                mp[s[right]]--;
            }
            while (count == 0) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minStart = left;
                }
                if (mp.count(s[left])) {
                    mp[s[left]]++;
                    if (mp[s[left]] > 0) count++;
                }
                left++;
            }
        }
        return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
    }
};

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-variable-shortest/minimum-window-substring/code/python/minimum-window-substring.py
def min_window(s, t):
    if len(s) < len(t):
        return ""
    mp = {}
    for c in t:
        mp[c] = mp.get(c, 0) + 1
    left = min_start = 0
    min_len = float('inf')
    count = len(t)
    for right in range(len(s)):
        if s[right] in mp:
            if mp[s[right]] > 0:
                count -= 1
            mp[s[right]] -= 1
        while count == 0:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                min_start = left
            if s[left] in mp:
                mp[s[left]] += 1
                if mp[s[left]] > 0:
                    count += 1
            left += 1
    return "" if min_len == float('inf') else s[min_start:min_start + min_len]